### ISSUE TITLE
SSL connection not working on port :9000

### ISSUE THREAD

**Issue by AI-Tax** [knadh/listmonk#692]
Hello,
 
I've installed listmonk successfully, everything works great - with one exception. I cannot post via API using ssl connection (https://). Also when I enter the port 9000 using http**s** I receive an error:
ERR_SSL_PROTOCOL_ERROR

While entering the website:9000 and not using http**s** everything is ok. Https connection works with subdomains and other address endings, it seems like it is only listmonk (port 9000) problem that is not working in ssl mode. Do you know how to enable that?

Best! 

PS: I am using nginx

**Comment by avanier** [comment]
@AI-Tax This is the expected behaviour. TLS is something you'll usually implement yourself in Nginx or whichever reverse-proxy you use to front your application. It's considered best practice to not implement TLS termination inside your application unless you have a specific needs to address. Since TLS is a security critical component in a lot of enterprises it's usually maintained separately, or globally for multiple applications.

If you want to get into it [Nginx's documentation](https://docs.nginx.com/nginx/admin-guide/security-controls/terminating-ssl-http/) on the matter is rather straightforward. _(As far as TLS goes.)_ They also have [a video](https://www.nginx.com/blog/nginx-https-101-ssl-basics-getting-started/) diving into the subject if reading isn't your jam.

**Comment by AI-Tax** [comment]
Hi @avanier many thanks for your support, I've just read the documents you sent - I've already implemented everything that is on the website. That's weird but the :9000 is not working in SSL. 

TLS seems to be working, in my nginx.conf I have a part:
	##
	# SSL Settings
	##

	ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

In my app's conf file I have nginx listening on port :443, I tried to listen also on port :9000 and then to forward that to http - but the solution did not work.

Maybe it could be useful for other webdev newbies if we try to find the solution here. If not, sorry for bothering you all and thanks for contributing.

**Comment by avanier** [comment]
@AI-Tax you've stepped in the realm of system administration more than webdev.

The key thing is that Listmonk itself will never answer to anything other than unencrypted HTTP traffic. You'll need another piece of software to speak HTTPS on the behalf of Listmonk.

The logic looks a bit like this:

```mermaid
flowchart LR
A[Client] -->|HTTPS:443| B(NGINX)
B -->|HTTP:9000| C[Listmonk]
B -->|HTTP:1234| D[Some other application]
```

The concept of TLS termination is that a proxy like Nginx will handle HTTPS connections on behalf of the applications it proxies. Nginx would talk to applications like Listmonk in cleartext HTTP, but will take care of doing the _"HTTPS"_ wrapping.

So overall you would configure Nginx to:

* Listen over 443 with TLS _(This is the part you did. :+1:)_
* Serve a [proxy block](https://docs.nginx.com/nginx/admin-guide/web-server/reverse-proxy/) to forward the decrypted requests to Listmonk _(This is probably the bit you were missing)_

   ```nginx
   # This assumes Listmonk is also running on the same machine as Nginx

   location / {
     proxy_set_header Host $host;
     proxy_set_header X-Real-IP $remote_addr;
     proxy_pass http://127.0.0.1:9000;
   }
   ```

**Comment by AI-Tax** [comment]
Many thanks for the clear explanation! I did it the way you mentioned before, that means I have two servers in nginx configuration file:

- the first one - listening on port 443 ssl, proxy pass to .sock file from gunicorn to let my flask app run,

- the second one - listening on port 9000, the part I have looks exactly like yours, with the exception that I am not using 127.0.0.1 but my domain name.

After your remarks I tried change the server name for the second server from the configuration file (like: mydomain.com/listmonk or subdomain.mydomain.com), listen for it on port 443 and to proxy_pass to http://mydomain.com:9000 - but it does not work.

I have simply run out of ideas how to solve it.
